internal/engine/upper_test.go

```diff
@@ -0,0 +1,143 @@
+package engine
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"sync"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/windmilleng/fsnotify"
+	"github.com/windmilleng/tilt/internal/logger"
+	"github.com/windmilleng/tilt/internal/model"
+	"github.com/windmilleng/tilt/internal/watch"
+)
+
+type fakeBuildAndDeployer struct {
+	startedServices []model.Service
+	cond            *sync.Cond
+}
+
+var _ BuildAndDeployer = &fakeBuildAndDeployer{}
+
+func (b *fakeBuildAndDeployer) BuildAndDeploy(ctx context.Context, service model.Service, token BuildToken) (BuildToken, error) {
+	b.startedServices = append(b.startedServices, service)
+	b.cond.L.Lock()
```

Comment:
you don't need to hold the lock to broadcast. see:
https://golang.org/pkg/sync/#Cond.Broadcast